### PR TITLE
Custome Pipelines

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR enables three use case:

## 1.) Download custom pipelines from a Hub model repo.

This requires the repo to have a `pipeline.py` file.

```python
from diffusers improt DiffusionPipeline

pipeline = DiffusionPipeline.from_pretrained("google/ddpm-cifar10-32", custom_pipeline="hf-internal-testing/diffusers-dummy-pipeline")

images, output_str = pipeline(num_inference_steps=2, output_type="np")
```

You can find the custom pipeline here: https://huggingface.co/hf-internal-testing/diffusers-dummy-pipeline/blob/main/pipeline.py

## 2.) We can also directly load community example pipelines from GitHub. 

E.g. the following works:

```python
from diffusers improt DiffusionPipeline
from transformers import CLIPFeatureExtractor, CLIPModel

clip_model_id = "laion/CLIP-ViT-B-32-laion2B-s34B-b79K"

feature_extractor = CLIPFeatureExtractor.from_pretrained(clip_model_id)
clip_model = CLIPModel.from_pretrained(clip_model_id)

pipeline = DiffusionPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-4",
    custom_pipeline="clip_guided_stable_diffusion",
    clip_model=clip_model,
    feature_extractor=feature_extractor,
)
pipeline = pipeline.to(torch_device)

# NOTE that `"CLIPGuidedStableDiffusion"` is not a class that is defined in the pypi package of th e library, but solely on the community examples folder of GitHub under:
# https://github.com/huggingface/diffusers/blob/main/examples/community/clip_guided_stable_diffusion.py
assert pipeline.__class__.__name__ == "CLIPGuidedStableDiffusion"

image = pipeline("a prompt", num_inference_steps=2, output_type="np").images[0]
```

This automatically imports community pipelines from: https://github.com/huggingface/diffusers/tree/main/examples/community
via GitHub.

## 3.) Local pipelines can also be loaded 
(see tests).

### CODE REVIEW

**Comment by pcuenca** [comment]
This looks great already!

> If costum_pipeline arg has a / it has to be on the Hub else it has to be on GitHub and such be an "official" community pipeline

Should we support local filesystem pipelines too? Of course they can simply load the code, but maybe this is convenient for some use cases.

**Comment by apolinario** [comment]
Excellent work @patrickvonplaten! The API design looks really good to me, I just have a few questions/comments

- How would it work when inside of a pipeline there's a library import that is not on diffusers requirements?
- Just to confirm: the current `DiffusionPipeline` function supports this, so I assume this design would also allow for passing the pipeline parameters downstream, is that right? (e.g.: 
```py
unet = UNet2DConditionModel.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="unet")
vae = AutoencoderKL.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="vae")
pipeline = DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4", custom_pipeline="clip_guided_stable_diffusion", vae=vae, unet=unet)
```
If so, then this is very exciting because we actually could do some meta-pipelines or pipeline stacking with the same model on VRAM
- I'm not so sure about the `/` being the separator between GitHub and HF Hub, plus the local loading challenge @pcuenca has mentioned. I think it is easy to use but not sure if it's easy to explain the difference. I wonder if adding different arguments for loading from HF Hub vs. Github couldn't make it easier to explain when documenting. Not a strong opinion on this one, brainstorming
- As this includes loading arbitrary code from the Hub, do we plan on having some automatic validation (e.g.: to guarantee this is indeed a diffusers pipeline and not just random code), could be something decentralised too (users reporting whether a pipeline works or is broken and we pass that information as a warning when user loads for e.g. if too many broken reports); this could also help with malicious code on pipelines for e.g.

**Comment by patrickvonplaten** [comment]
Very good questions!  Answering in line.

> Excellent work @patrickvonplaten! The API design looks really good to me, I just have a few questions/comments
> 
> * How would it work when inside of a pipeline there's a library import that is not on diffusers requirements?

Yes, I've thought about this - it makes a lot of sense to me to provide this, but thought maybe in a v2 version, but generally it should already work out of the box as long as the library import is installed. In a v2 version, we could do some more fancy checking where we show a nice error message if a library import is not installed

> * Just to confirm: the current `DiffusionPipeline` function supports this, so I assume this design would also allow for passing the pipeline parameters downstream, is that right? (e.g.:
> 
> ```python
> unet = UNet2DConditionModel.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="unet")
> vae = AutoencoderKL.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="vae")
> pipeline = DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4", custom_pipeline="clip_guided_stable_diffusion", vae=vae, unet=unet)
> ```

Yes exactly, the plan is to make exactly this API work now!

> 
> If so, then this is very exciting because we actually could do some meta-pipelines or pipeline stacking with the same model on VRAM
> 
> * I'm not so sure about the `/` being the separator between GitHub and HF Hub, plus the local loading challenge @pcuenca has mentioned. I think it is easy to use but not sure if it's easy to explain the difference. I wonder if adding different arguments for loading from HF Hub vs. Github couldn't make it easier to explain when documenting. Not a strong opinion on this one, brainstorming

Good feedback! It's mainly borrowed from how we differentiated "canonical" from non-canonical datasets in the earlier days of `datasets`. Maybe in the beginning we could take this slightly magical approach. Local folders should always be suffixed by a `./` (that's already the case in `transformers` and `datasets` so I think this should be fine

> * As this includes loading arbitrary code from the Hub, do we plan on having some automatic validation (e.g.: to guarantee this is indeed a diffusers pipeline and not just random code), could be something decentralised too (users reporting whether a pipeline works or is broken and we pass that information as a warning when user loads for e.g. if too many broken reports); this could also help with malicious code on pipelines for e.g.

Yeah also good question - for now I'd just put a red flag in the docs under the `custom_pipeline` argument stating that "unofficial" code is run. I'd also say that we'll see down the road if we need more safety mechanisms such as a "`trust_remote_code`" flag.

**Comment by patrickvonplaten** [comment]
> This looks great already!
> 
> > If costum_pipeline arg has a / it has to be on the Hub else it has to be on GitHub and such be an "official" community pipeline
> 
> Should we support local filesystem pipelines too? Of course they can simply load the code, but maybe this is convenient for some use cases.

Yes trying to get this to work as well!

**Comment by patil-suraj** [comment]
The design looks really good, pretty much the same comments as Pedro and Poli.

One question, what's the naming convention, for example if `custom_pipeline="some_user/clip_guided_stable_diffusion"`
then how should the repo look like ? Both the repo and pipeline file should be named `clip_guided_stable_diffusion` ?

Also, would it possible to provide a custom pipeline along with a custom checkpoint or do we need to create a different repo for the pipeline ?

**Comment by patrickvonplaten** [comment]
> The design looks really good, pretty much the same comments as Pedro and Poli.
> 
> One question, what's the naming convention, for example if `custom_pipeline="some_user/clip_guided_stable_diffusion"` then how should the repo look like ? Both the repo and pipeline file should be named `clip_guided_stable_diffusion` ?

For now the convention is - if it's not a GitHub community example, then the file should be called `pipeline.py` 

> 
> Also, would it possible to provide a custom pipeline along with a custom checkpoint or do we need to create a different repo for the pipeline ?

Sure that should already work, one can just pass the costum checkpoint as `pretrained_model_or_path` and it should work

File: tests/fixtures/costum_pipeline/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            # do x_t -> x_t-1
+            image = self.scheduler.step(model_output, t, image, eta).prev_sample
+
+        image = (image / 2 + 0.5).clamp(0, 1)
+        image = image.cpu().permute(0, 2, 3, 1).numpy()
+        if output_type == "pil":
+            image = self.numpy_to_pil(image)
+
+        if not return_dict:
+            return (image,)
```
Comment (pcuenca) [inline tests/fixtures/costum_pipeline/pipeline.py:100]: Extremely minor: add the string as a return type here in case we want to change the test later?
```suggestion
            return (image,), "This is a local test"
```

**Review by pcuenca — APPROVED** [review]
Amazing work, love the tests. I just suggested some typo fixes.

File: docs/source/using-diffusers/custom_pipelines.mdx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        # do x_t -> x_t-1
+        image = self.scheduler.step(model_output, t, image, eta).prev_sample
+
+    image = (image / 2 + 0.5).clamp(0, 1)
+    image = image.cpu().permute(0, 2, 3, 1).numpy()
+
+    return image
+```
+
+Now you can upload this short file under the name `pipeline.py` in your prefered model repository.
```
Comment (apolinario) [inline docs/source/using-diffusers/custom_pipelines.mdx:114]:
```suggestion
Now you can upload this short file under the name `pipeline.py` in your preferred [model repository](https://huggingface.co/docs/hub/models-uploading). For Stable Diffusion pipelines, you may also [join the community organisation for shared pipelines](https://huggingface.co/organizations/sd-diffusers-pipelines-library/share/BUPyDUuHcciGTOKaExlqtfFcyCZsVFdrjr) to upload yours.
```

File: docs/source/using-diffusers/custom_pipelines.mdx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+Finally, we can load the custom pipeline by passing the model repository name, *e.g.* `patrickvonplaten/my_custom_pipeline` along side the model repository from where we want to load the `unet` and `scheduler` components.
```
Comment (apolinario) [inline docs/source/using-diffusers/custom_pipelines.mdx:115]:
```suggestion
Finally, we can load the custom pipeline by passing the model repository name, *e.g.* `sd-diffusers-pipelines-library/my_custom_pipeline` alongside the model repository from where we want to load the `unet` and `scheduler` components.
```

Comment (apolinario) [reply docs/source/using-diffusers/custom_pipelines.mdx:115]: Have the community pipeline org as the example

File: src/diffusers/dynamic_modules_utils.py [hunk 4]
```diff
@@ -208,24 +240,43 @@ def get_cached_module_file(
...
+
     module_file_or_url = os.path.join(pretrained_model_name_or_path, module_file)
-    submodule = "local"
 
     if os.path.isfile(module_file_or_url):
         resolved_module_file = module_file_or_url
+        submodule = "local"
+    elif pretrained_model_name_or_path.count("/") == 0:
+        # community pipeline on GitHub
+        github_url = COMMUNITY_PIPELINES_URL.format(pipeline=pretrained_model_name_or_path)
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/dynamic_modules_utils.py:251]: yes def in a v2 version if this feature is used this could make sense

File: docs/source/using-diffusers/custom_pipelines.mdx (hunk 3 above)
Comment (patrickvonplaten) [reply docs/source/using-diffusers/custom_pipelines.mdx:115]: Can you invite me to the repo so that I can transfer: https://huggingface.co/patrickvonplaten/my_custom_pipeline then?
